libsql-sqlite3/src/vectordiskann.c

```diff
@@ -0,0 +1,865 @@
+      return "float32";
+    default:
+      return NULL;
+  }
+}
+
+static int diskAnnFromVectorType(const char *zVectorType){
+  if( strcmp(zVectorType, "float32")==0 ){
+    return VECTOR_TYPE_FLOAT32;
+  }
+  return -1;
+}
+
+static const char *diskAnnToDistanceOps(int nDistanceFunc){
+  switch( nDistanceFunc ){
+    case VECTOR_DISTANCE_COS:
+      return "cosine";
+    default:
+      return NULL;
+  }
+}
+
+static int diskannFromDistanceOps(const char *zDistanceOps){
+  if( strcmp(zDistanceOps, "cosine")==0 ){
+    return VECTOR_DISTANCE_COS;
+  }
+  return -1;
+}
+
+int diskAnnCreateIndex(
+  sqlite3 *db,
+  const char *zIdxName,
+  unsigned int nDims,
+  unsigned int nDistanceFunc
+){
+  const char *zDistanceOps;
+  DiskAnnIndex *pIndex;
+  sqlite3_stmt *pStmt;
+  int rc;
+
+  zDistanceOps = diskAnnToDistanceOps(nDistanceFunc);
+  if( zDistanceOps==NULL ){
+    return SQLITE_ERROR;
+  }
+  pIndex = sqlite3_malloc(sizeof(DiskAnnIndex));
+  if( pIndex == NULL ){
+    return SQLITE_NOMEM;
+  }
+  pIndex->db = db;
+  pIndex->zDb = strdup(db->aDb[0].zDbSName);
+  pIndex->zName = strdup(zIdxName);
+  pIndex->zShadow = sqlite3MPrintf(db, "%s_shadow", zIdxName);
+  pIndex->nDistanceFunc = nDistanceFunc;
+  pIndex->nBlockSize = DISKANN_BLOCK_SIZE >> DISKANN_BLOCK_SIZE_SHIFT;
+  pIndex->nVectorType = VECTOR_TYPE_FLOAT32;
+  pIndex->nVectorDims = nDims;
+
+  static const char zInsertSql[] =
+    "INSERT INTO libsql_vector_index "
+    "(type, name, vector_type, block_size, dims, distance_ops)"
+    "VALUES "
+    "(?, ?, ?, ?, ?, ?)";
+    
+  rc = sqlite3_prepare_v2(db, zInsertSql, -1, &pStmt, 0);
+  if( rc!=SQLITE_OK ){
+    return rc;
+  }
+  sqlite3_bind_text(pStmt, 1, "diskann", -1, SQLITE_STATIC);
+  sqlite3_bind_text(pStmt, 2, zIdxName, -1, SQLITE_STATIC);
+  sqlite3_bind_text(pStmt, 3, diskAnnToVectorType(pIndex->nVectorType), -1, SQLITE_STATIC);
+  sqlite3_bind_int (pStmt, 4, DISKANN_BLOCK_SIZE >> DISKANN_BLOCK_SIZE_SHIFT);
+  sqlite3_bind_int (pStmt, 5, nDims);
+  sqlite3_bind_text(pStmt, 6, diskAnnToDistanceOps(pIndex->nDistanceFunc), -1, SQLITE_STATIC);
+  rc = sqlite3_step(pStmt);
+  sqlite3_finalize(pStmt);
+  diskAnnCloseIndex(pIndex);
+  return rc;
+}
+
+int diskAnnOpenIndex(
+  sqlite3 *db,                    /* Database connection */
+  const char *zIdxName,           /* Index name */
+  DiskAnnIndex **ppIndex          /* OUT: Index */
+){
+  DiskAnnIndex *pIndex;
+  sqlite3_stmt *pStmt;
+  int rc = SQLITE_OK;
+
+  static const char zInsertSql[] =
+    "SELECT vector_type, block_size, dims, distance_ops "
+    "FROM libsql_vector_index "
+    "WHERE type = ? AND name = ?";
+  rc = sqlite3_prepare_v2(db, zInsertSql, -1, &pStmt, 0);
+  if( rc!=SQLITE_OK ){
+    printf("Failed to prepare statement\n");
+    return rc;
+  }
+  sqlite3_bind_text(pStmt, 1, "diskann", -1, SQLITE_STATIC);
+  sqlite3_bind_text(pStmt, 2, zIdxName, -1, SQLITE_STATIC);
+  if( sqlite3_step(pStmt)!=SQLITE_ROW ){
+    printf("Failed to step\n");
```

Comment:
My insert query fails at this part when I have an existing unique index on my table.

I tried to dig deeper but only got as far as realizing that this SQL is executed on my unique index as well, resulting in an error because the SQL returned an `SQLITE_DONE` due to no results.

Minimal reproduction steps if it helps:
```
CREATE TABLE movies (title TEXT UNIQUE, embedding FLOAT32(3));
CREATE INDEX movies_idx USING diskann_cosine_ops ON movies (embedding);
INSERT INTO movies (title, embedding) VALUES ('a', vector('[1,2,3]'));
```

Hoping for a fix so that I can use it with my table with unique indexes, but cheers for the vector search support!

Reply:
@xiankai https://github.com/tursodatabase/libsql/pull/1402/commits/acc21621c96c56dcb31548b3b23fd595bfa4ea1d should fix that issue.